%\VignetteIndexEntry{GOsim}
%\VignetteKeyword{GOsim}
%\VignettePackage{GOsim}
\documentclass[12pt,a4paper]{article}
\usepackage[round]{natbib}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{hyperref}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\textwidth=6.2in
\textheight=8.5in
%\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}

\usepackage{/usr/share/R/share/texmf/Sweave}
\begin{document}
\author{Holger FrÃ¶hlich}

\title{The \emph{GOSim} package}
\date{\today}
\maketitle

\section{Introduction}

The Gene Ontology (GO) has become one of the most widespread systems for systematically annotating gene products within the bioinformatics community and is developed by the Gene Ontology Consortium \cite{GOConsortium04}. It is specifically intended for describing gene products with a controlled and structured vocabulary. GO terms are part of a Directed Acyclic Graph (DAG), covering three orthogonal taxonomies or "aspects": \emph{molecular function, biological process} and \emph{cellular component}. Two different kinds of relationship between GO terms exist: the "is-a" relationship and the "part-of" relationship.
Providing a standard vocabulary across any biological resources, the GO enables researchers to use this information for automated data analysis.

The \emph{GOSim} package provides the researcher with various information theoretic similarity concepts for GO terms \cite{Resnik95, Resnik99, Lin98, Jiang98, Lord03, Couto2003FuSSiMeg, Couto2005GraSM}. It additionally implements different methods for computing functional similarities between gene products based on the similarties between the associated GO terms. This can, for instances, be used for clustering genes according to their biological function \cite{FroeGO05, FroeSpeerGOKer06} and thus may help to get a better understanding of the biological aspects covered by a set of genes.

\section{Usage of \emph{GOSim}}

To elucidate the usage of \emph{GOSim} we show an example workflow and explain the employed similarity concepts. We create  a character vector of Entrez gene IDs:
\begin{Schunk}
\begin{Sinput}
> library(GOSim)
> genes = c("207", "208", "596", "901", "780", "3169", "9518", 
+     "2852", "26353", "8614", "7494")
\end{Sinput}
\end{Schunk}
Next we investigate the GO annotation within the current ontology (which is \emph{biological process} by default):
\begin{Schunk}
\begin{Sinput}
> getGOInfo(genes)
\end{Sinput}
\end{Schunk}

\subsection{Term Similarities}

Let us examine the similarity of the GO terms for genes "8614" and "2852" in greater detail:
\begin{Schunk}
\begin{Sinput}
> getTermSim(c("GO:0007166", "GO:0007267", "GO:0007584", "GO:0007165", 
+     "GO:0007186"), method = "Resnik", verbose = FALSE)
\end{Sinput}
\begin{Soutput}
           GO:0007166 GO:0007267 GO:0007584 GO:0007165 GO:0007186
GO:0007166  1.0000000  0.3225697  0.3225697  0.3317104  0.3787826
GO:0007267  0.3225697  1.0000000  0.3225697  0.3225697  0.3225697
GO:0007584  0.3225697  0.3225697  1.0000000  0.3225697  0.3225697
GO:0007165  0.3317104  0.3225697  0.3225697  1.0000000  0.3317104
GO:0007186  0.3787826  0.3225697  0.3225697  0.3317104  1.0000000
\end{Soutput}
\end{Schunk}
This calculates Resnik's pairwise similarity between GO terms \cite{Resnik95,Resnik99}:
\begin{equation}
sim(t,t') = IC_{ms}(t,t') := \max_{\hat{t}\in Pa(t,t')} IC(\hat{t})\label{eq:Resnik}
\end{equation}
Here $Pa(t,t')$ denotes the set of all common ancestors of GO terms $t$ and $t'$, while $IC(t)$  denotes the information content of term $t$. It is defined as (e.g. \cite{Lord03})
\begin{equation}
IC(\hat{t}) = -\log P(\hat{t})
\end{equation}
i.e. as the negative logarithm of the probability of observing $t$. The information content of each GO term is already precomputed for each ontology based on the empirical observation, how many times a specific GO term or any of its direct or indirect offsprings appear in the annotation of the GO with gene products. The association between gene products and GO identifiers is reported regularily by the NCBI.

\begin{Schunk}
\begin{Sinput}
> data("ICsBPall")
> IC[c("GO:0007166", "GO:0007267", "GO:0007584", "GO:0007165", 
+     "GO:0007186")]
\end{Sinput}
\begin{Soutput}
GO:0007166 GO:0007267 GO:0007584 GO:0007165 GO:0007186 
  5.540203   6.900127  10.776192   4.851709   5.945838 
\end{Soutput}
\end{Schunk}
This loads the information contents of all GO terms within "biological process". Likewise, the data files {\tt ICsMFall} and {\tt ICsCCall} contain the information contents of all GO terms within "molecular function" and "cellular component". If only GO terms having evidence codes "IMP" (inferred from mutant phenotype), "IGI", (inferred from genetic interaction), "IDA" (inferred from direct assay), "IEP" (inferred from expression pattern) or "IPI" (inferred from physical interaction) are wanted, one can use the data files {\tt ICsBPIMP\_IGI\_IDA\_IEP\_IPI}, {\tt ICsMFIMP\_IGI\_IDA\_IEP\_IPI} and {\tt ICsCCIMP\_IGI\_IDA\_IEP\_IPI}, respectively. The information contents for GO terms filtered with respect to different evidence codes must be calculated explicitely using the function {\tt calcICs}. Please refer to the manual pages for details.

For the similarity computation in (Eq.: \ref{eq:Resnik}) normalized information contents are used, which are obtained by dividing the raw information contents by its maximal value:
\begin{Schunk}
\begin{Sinput}
> IC[c("GO:0007166", "GO:0007267", "GO:0007584", "GO:0007165", 
+     "GO:0007186")]/max(IC[IC != Inf])
\end{Sinput}
\begin{Soutput}
GO:0007166 GO:0007267 GO:0007584 GO:0007165 GO:0007186 
 0.3787826  0.4717604  0.7367662  0.3317104  0.4065158 
\end{Soutput}
\end{Schunk}

To continue our example from above, let us also calculate Jiang and Conrath's pairwise similarity between GO terms, which is the default, for compairson reasons \cite{Jiang98}:
\begin{Schunk}
\begin{Sinput}
> getTermSim(c("GO:0007166", "GO:0007267", "GO:0007584", "GO:0007165", 
+     "GO:0007186"), verbose = FALSE)
\end{Sinput}
\begin{Soutput}
           GO:0007166 GO:0007267 GO:0007584 GO:0007165 GO:0007186
GO:0007166  1.0000000  0.7945964  0.5295906  0.9529278  0.9722668
GO:0007267  0.7945964  1.0000000  0.4366129  0.8416687  0.7668633
GO:0007584  0.5295906  0.4366129  1.0000000  0.5766628  0.5018574
GO:0007165  0.9529278  0.8416687  0.5766628  1.0000000  0.9251946
GO:0007186  0.9722668  0.7668633  0.5018574  0.9251946  1.0000000
\end{Soutput}
\end{Schunk}
Jiang and Conrath's similarity measure is defined as
\begin{equation}
sim(t,t') = 1 - \min(1, IC(t) - 2IC_{ms}(t,t') + IC(t'))
\end{equation}
i.e. the similarity between $t$ and $t'$ is 0, if their normalized distance is at least 1.

Likewise, we can also compute Lin's pairwise similarity between GO terms \cite{Lin98}:
\begin{Schunk}
\begin{Sinput}
> getTermSim(c("GO:0007166", "GO:0007267", "GO:0007584", "GO:0007165", 
+     "GO:0007186"), method = "Lin", verbose = FALSE)
\end{Sinput}
\begin{Soutput}
           GO:0007166 GO:0007267 GO:0007584 GO:0007165 GO:0007186
GO:0007166  1.0000000  0.7585030  0.5783157  0.9337471  0.9646845
GO:0007267  0.7585030  1.0000000  0.5338231  0.8029407  0.7345519
GO:0007584  0.5783157  0.5338231  1.0000000  0.6037937  0.5642872
GO:0007165  0.9337471  0.8029407  0.6037937  1.0000000  0.8986687
GO:0007186  0.9646845  0.7345519  0.5642872  0.8986687  1.0000000
\end{Soutput}
\end{Schunk}
It is defined as:
\begin{equation}
sim(t,t') = \frac{2IC_{ms}(t,t')}{IC(t) + IC(t')}\label{eq:Lin}
\end{equation}

Resnik's, Jiang-Conraths's and Lin's term similarities all refer to $IC_{ms}(t,t')$, the information content of the minimum subsumer of $t$ and $t'$, i.e. of the lowest common ancestor in the hierarchy. For illustration let us plot the GO graph with leaves GO:0007166 and GO:0007267 and let us compute their minimum subsumer (see Fig. \ref{Fig:GOPlot}):
\begin{Schunk}
\begin{Sinput}
> library(Rgraphviz)